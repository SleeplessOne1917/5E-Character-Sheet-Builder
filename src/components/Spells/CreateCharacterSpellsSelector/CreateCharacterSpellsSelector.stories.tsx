import { ComponentMeta, ComponentStory } from '@storybook/react';

import CreateCharacterSpellsSelector from './CreateCharacterSpellsSelector';
import { Provider } from 'react-redux';
import { getTestStore } from '../../../redux/store';

export default {
	title: 'Components/Spells/CreateCharacterSpellsSelector',
	component: CreateCharacterSpellsSelector,
	args: {
		choose: 5,
		spells: [
			{
				id: 'true-strike',
				name: 'True Strike',
				level: 0,
				school: {
					name: 'Divination',
					id: 'divination'
				}
			},
			{
				id: 'light',
				name: 'Light',
				level: 0,
				school: {
					name: 'Evocation',
					id: 'evocation'
				}
			},
			{
				id: 'mending',
				name: 'Mending',
				level: 0,
				school: {
					name: 'Transmutation',
					id: 'transmutation'
				}
			},
			{
				id: 'message',
				name: 'Message',
				level: 0,
				school: {
					name: 'Transmutation',
					id: 'transmutation'
				}
			},
			{
				id: 'vicious-mockery',
				name: 'Vicious Mockery',
				level: 0,
				school: {
					name: 'Enchantment',
					id: 'enchantment'
				}
			},
			{
				id: 'minor-illusion',
				name: 'Minor Illusion',
				level: 0,
				school: {
					name: 'Illusion',
					id: 'illusion'
				}
			},
			{
				id: 'dancing-lights',
				name: 'Dancing Lights',
				level: 0,
				school: {
					name: 'Evocation',
					id: 'evocation'
				}
			}
		]
	},
	decorators: [
		Story => (
			<Provider store={getTestStore()}>
				<Story />
			</Provider>
		)
	]
} as ComponentMeta<typeof CreateCharacterSpellsSelector>;

const Template: ComponentStory<typeof CreateCharacterSpellsSelector> = args => (
	<CreateCharacterSpellsSelector {...args} />
);

export const Default = Template.bind({});
