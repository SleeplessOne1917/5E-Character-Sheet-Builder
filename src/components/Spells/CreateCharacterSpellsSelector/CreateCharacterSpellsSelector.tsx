'use client';

import {
	addClassSpell,
	removeClassSpell
} from '../../../redux/features/classInfo';
import { addSpell, removeSpell } from '../../../redux/features/spellcasting';
import { useAppDispatch, useAppSelector } from '../../../hooks/reduxHooks';
import { useCallback, useMemo, useState } from 'react';

import { SpellItem } from '../../../types/characterSheetBuilderAPI';
import SpellsSelector from '../SpellsSelector/SpellsSelector';

type CreateCharacterSpellsSelectorProps = {
	spells: SpellItem[];
	choose: number;
};

const CreateCharacterSpellsSelector = ({
	spells,
	choose
}: CreateCharacterSpellsSelectorProps) => {
	const classSpells = useAppSelector(
		state => state.editingCharacter.classInfo.spells
	);
	const highestSlotLevel = useAppSelector(
		state => state.editingCharacter.spellcasting.highestSlotLevel
	);
	const levels = useMemo(() => {
		const levels: number[] = [];
		if (
			!spells.reduce<{ level: number | null; isSameLevel: boolean }>(
				(acc, cur) => {
					if (!acc.level) {
						return {
							...acc,
							level: cur.level
						};
					} else {
						return {
							...acc,
							isSameLevel: cur.level === acc.level
						};
					}
				},
				{
					level: null,
					isSameLevel: true
				}
			).isSameLevel
		) {
			for (let i = 1; i <= highestSlotLevel; ++i) {
				levels.push(i);
			}
		}

		return levels;
	}, [highestSlotLevel, spells]);

	const dispatch = useAppDispatch();

	const traitSpellsByTrait = useAppSelector(
		state => state.editingCharacter.raceInfo.selectedTraitSpells
	);

	const traitSpells = useMemo(
		() => Object.values(traitSpellsByTrait).flatMap(value => value),
		[traitSpellsByTrait]
	);

	const subclassSpells = useAppSelector(
		state => state.editingCharacter.classInfo.subclass?.spells
	);

	const getInitialSelectedSpells = useCallback(
		() =>
			spells
				.filter(spell => classSpells?.some(cs => cs.id === spell.id))
				.map(({ id }) => id),
		[spells, classSpells]
	);

	const [selectedSpells, setSelectedSpells] = useState(
		getInitialSelectedSpells()
	);

	const handleAdd = useCallback(
		(spell: SpellItem) => {
			dispatch(addClassSpell(spell));
			dispatch(addSpell(spell));
			setSelectedSpells(prev => [...prev, spell.id]);
		},
		[dispatch]
	);

	const handleRemove = useCallback(
		(spell: SpellItem) => {
			dispatch(removeClassSpell(spell.id));
			dispatch(removeSpell(spell.id));
			setSelectedSpells(prev => prev.filter(s => s !== spell.id));
		},
		[dispatch]
	);

	const filterSpell = useCallback(
		(spell: SpellItem) =>
			!traitSpells.some(ts => ts.id === spell.id) &&
			!(subclassSpells && subclassSpells.some(s => s.spell.index === spell.id)),
		[traitSpells, subclassSpells]
	);

	const numberOfSelectedSpells = selectedSpells.filter(
		spell => spell !== 'blank'
	).length;

	return (
		<SpellsSelector
			isSelected={numberOfSelectedSpells === choose}
			selectedSpells={selectedSpells}
			spells={spells}
			levels={levels}
			label={`${numberOfSelectedSpells}/${choose} ${
				spells.some(({ level }) => level > 0) ? 'spell' : 'cantrip'
			}s selected`}
			filterSpell={filterSpell}
			onAdd={handleAdd}
			onRemove={handleRemove}
		/>
	);
};

export default CreateCharacterSpellsSelector;
