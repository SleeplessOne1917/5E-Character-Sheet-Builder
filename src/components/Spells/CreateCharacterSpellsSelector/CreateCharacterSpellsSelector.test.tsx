import * as stories from './CreateCharacterSpellsSelector.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/testing-react';

const { Default } = composeStories(stories);

window.matchMedia = jest.fn().mockImplementation(() => ({
	matches: true,
	addEventListener: jest.fn(),
	removeEventListener: jest.fn()
}));

describe('renders correctly', () => {
	it('default', () => {
		render(<Default />);

		expect(screen.getByTestId('spells-selector')).toMatchSnapshot();
	});
});
