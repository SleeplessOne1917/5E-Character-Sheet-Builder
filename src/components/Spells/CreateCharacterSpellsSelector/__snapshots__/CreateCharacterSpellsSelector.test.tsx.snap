// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly default 1`] = `
<div
  class="container"
  data-testid="spells-selector"
>
  <div
    class="text"
  >
    0/5 cantrips selected
  </div>
  <div
    class="filter-bar"
  >
    <div
      class="search-container"
    >
      <input
        aria-label="Search cantrips"
        class="search"
        placeholder="Search cantrips"
        value=""
      />
      <svg
        aria-hidden="true"
        class="search-icon"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M9 3.5a5.5 5.5 0 100 11 5.5 5.5 0 000-11zM2 9a7 7 0 1112.452 4.391l3.328 3.329a.75.75 0 11-1.06 1.06l-3.329-3.328A7 7 0 012 9z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="selects"
    >
      <div
        class="select-container"
      >
        <div
          class="list-box-and-error-container"
          data-testid="select"
        >
          <label
            id="school-select-label"
            style="font-size: 1.5rem;"
          >
            School
          </label>
          <div
            aria-labelledby="school-select-label"
            class="list-box-container"
            style="font-size: 1.2rem;"
          >
            <button
              aria-expanded="false"
              aria-haspopup="listbox"
              class="button"
              data-testid="select-button"
              type="button"
            >
              —
              <svg
                aria-hidden="true"
                class="icon"
                fill="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
            <ul
              aria-activedescendant="—"
              class="list"
              data-testid="select-list"
              role="listbox"
              style="display: none;"
              tabindex="-1"
            >
              <li
                aria-selected="true"
                class="item selected"
                data-testid="data-blank"
                role="option"
                tabindex="0"
              >
                —
              </li>
              <li
                aria-selected="false"
                class="item"
                data-testid="data-divination"
                role="option"
                tabindex="0"
              >
                Divination
              </li>
              <li
                aria-selected="false"
                class="item"
                data-testid="data-evocation"
                role="option"
                tabindex="0"
              >
                Evocation
              </li>
              <li
                aria-selected="false"
                class="item"
                data-testid="data-transmutation"
                role="option"
                tabindex="0"
              >
                Transmutation
              </li>
              <li
                aria-selected="false"
                class="item"
                data-testid="data-enchantment"
                role="option"
                tabindex="0"
              >
                Enchantment
              </li>
              <li
                aria-selected="false"
                class="item"
                data-testid="data-illusion"
                role="option"
                tabindex="0"
              >
                Illusion
              </li>
            </ul>
          </div>
        </div>
      </div>
      <label
        class="label"
        data-testid="checkbox"
      >
        <div
          aria-checked="false"
          class="checkbox"
          role="checkbox"
          tabindex="0"
        />
        Only Show Selected
      </label>
    </div>
  </div>
  <div
    class="spells-container"
  >
    <div
      class="selector"
      data-testid="spell-selector"
    >
      <div
        aria-label="True Strike"
        class="selector-label"
        role="button"
        tabindex="0"
      >
        <div
          class="label-name-container"
        >
          <svg
            class="label-icon"
          >
            <use
              xlink:href="/Icons.svg#divination"
            />
          </svg>
          True Strike
        </div>
        <svg
          aria-hidden="true"
          class="label-icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="selector-display"
        style="display: none;"
      >
        <div
          class="summary"
        >
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Level
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Casting time
            </span>
            : 
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Duration
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Range
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              School
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Components
            </span>
            : 
          </div>
        </div>
        <div
          class="description"
        >
          <div
            style="display: flex; align-items: center; justify-content: center;"
          >
            <div
              class="container"
            >
              <svg
                class="icon"
              >
                <use
                  xlink:href="/Icons.svg#spinner"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="other"
        />
        <div
          class="buttons"
        >
          <button
            class="button positive"
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Add
          </button>
          <button
            class="button negative"
            disabled=""
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
    <div
      class="selector"
      data-testid="spell-selector"
    >
      <div
        aria-label="Light"
        class="selector-label"
        role="button"
        tabindex="0"
      >
        <div
          class="label-name-container"
        >
          <svg
            class="label-icon"
          >
            <use
              xlink:href="/Icons.svg#evocation"
            />
          </svg>
          Light
        </div>
        <svg
          aria-hidden="true"
          class="label-icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="selector-display"
        style="display: none;"
      >
        <div
          class="summary"
        >
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Level
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Casting time
            </span>
            : 
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Duration
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Range
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              School
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Components
            </span>
            : 
          </div>
        </div>
        <div
          class="description"
        >
          <div
            style="display: flex; align-items: center; justify-content: center;"
          >
            <div
              class="container"
            >
              <svg
                class="icon"
              >
                <use
                  xlink:href="/Icons.svg#spinner"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="other"
        />
        <div
          class="buttons"
        >
          <button
            class="button positive"
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Add
          </button>
          <button
            class="button negative"
            disabled=""
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
    <div
      class="selector"
      data-testid="spell-selector"
    >
      <div
        aria-label="Mending"
        class="selector-label"
        role="button"
        tabindex="0"
      >
        <div
          class="label-name-container"
        >
          <svg
            class="label-icon"
          >
            <use
              xlink:href="/Icons.svg#transmutation"
            />
          </svg>
          Mending
        </div>
        <svg
          aria-hidden="true"
          class="label-icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="selector-display"
        style="display: none;"
      >
        <div
          class="summary"
        >
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Level
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Casting time
            </span>
            : 
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Duration
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Range
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              School
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Components
            </span>
            : 
          </div>
        </div>
        <div
          class="description"
        >
          <div
            style="display: flex; align-items: center; justify-content: center;"
          >
            <div
              class="container"
            >
              <svg
                class="icon"
              >
                <use
                  xlink:href="/Icons.svg#spinner"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="other"
        />
        <div
          class="buttons"
        >
          <button
            class="button positive"
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Add
          </button>
          <button
            class="button negative"
            disabled=""
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
    <div
      class="selector"
      data-testid="spell-selector"
    >
      <div
        aria-label="Message"
        class="selector-label"
        role="button"
        tabindex="0"
      >
        <div
          class="label-name-container"
        >
          <svg
            class="label-icon"
          >
            <use
              xlink:href="/Icons.svg#transmutation"
            />
          </svg>
          Message
        </div>
        <svg
          aria-hidden="true"
          class="label-icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="selector-display"
        style="display: none;"
      >
        <div
          class="summary"
        >
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Level
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Casting time
            </span>
            : 
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Duration
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Range
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              School
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Components
            </span>
            : 
          </div>
        </div>
        <div
          class="description"
        >
          <div
            style="display: flex; align-items: center; justify-content: center;"
          >
            <div
              class="container"
            >
              <svg
                class="icon"
              >
                <use
                  xlink:href="/Icons.svg#spinner"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="other"
        />
        <div
          class="buttons"
        >
          <button
            class="button positive"
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Add
          </button>
          <button
            class="button negative"
            disabled=""
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
    <div
      class="selector"
      data-testid="spell-selector"
    >
      <div
        aria-label="Vicious Mockery"
        class="selector-label"
        role="button"
        tabindex="0"
      >
        <div
          class="label-name-container"
        >
          <svg
            class="label-icon"
          >
            <use
              xlink:href="/Icons.svg#enchantment"
            />
          </svg>
          Vicious Mockery
        </div>
        <svg
          aria-hidden="true"
          class="label-icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="selector-display"
        style="display: none;"
      >
        <div
          class="summary"
        >
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Level
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Casting time
            </span>
            : 
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Duration
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Range
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              School
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Components
            </span>
            : 
          </div>
        </div>
        <div
          class="description"
        >
          <div
            style="display: flex; align-items: center; justify-content: center;"
          >
            <div
              class="container"
            >
              <svg
                class="icon"
              >
                <use
                  xlink:href="/Icons.svg#spinner"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="other"
        />
        <div
          class="buttons"
        >
          <button
            class="button positive"
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Add
          </button>
          <button
            class="button negative"
            disabled=""
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
    <div
      class="selector"
      data-testid="spell-selector"
    >
      <div
        aria-label="Minor Illusion"
        class="selector-label"
        role="button"
        tabindex="0"
      >
        <div
          class="label-name-container"
        >
          <svg
            class="label-icon"
          >
            <use
              xlink:href="/Icons.svg#illusion"
            />
          </svg>
          Minor Illusion
        </div>
        <svg
          aria-hidden="true"
          class="label-icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="selector-display"
        style="display: none;"
      >
        <div
          class="summary"
        >
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Level
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Casting time
            </span>
            : 
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Duration
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Range
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              School
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Components
            </span>
            : 
          </div>
        </div>
        <div
          class="description"
        >
          <div
            style="display: flex; align-items: center; justify-content: center;"
          >
            <div
              class="container"
            >
              <svg
                class="icon"
              >
                <use
                  xlink:href="/Icons.svg#spinner"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="other"
        />
        <div
          class="buttons"
        >
          <button
            class="button positive"
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Add
          </button>
          <button
            class="button negative"
            disabled=""
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
    <div
      class="selector"
      data-testid="spell-selector"
    >
      <div
        aria-label="Dancing Lights"
        class="selector-label"
        role="button"
        tabindex="0"
      >
        <div
          class="label-name-container"
        >
          <svg
            class="label-icon"
          >
            <use
              xlink:href="/Icons.svg#evocation"
            />
          </svg>
          Dancing Lights
        </div>
        <svg
          aria-hidden="true"
          class="label-icon"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="selector-display"
        style="display: none;"
      >
        <div
          class="summary"
        >
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Level
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Casting time
            </span>
            : 
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Duration
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Range
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              School
            </span>
            :
             
          </div>
          <div
            class="summary-item"
          >
            <span
              class="summary-item-label"
            >
              Components
            </span>
            : 
          </div>
        </div>
        <div
          class="description"
        >
          <div
            style="display: flex; align-items: center; justify-content: center;"
          >
            <div
              class="container"
            >
              <svg
                class="icon"
              >
                <use
                  xlink:href="/Icons.svg#spinner"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="other"
        />
        <div
          class="buttons"
        >
          <button
            class="button positive"
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Add
          </button>
          <button
            class="button negative"
            disabled=""
            style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
